vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/TextField.java

```diff
@@ -40,12 +47,13 @@
 public class TextField extends TextFieldBase<TextField, String>
         implements HasAllowedCharPattern, HasThemeVariant<TextFieldVariant> {
 
-    private boolean isConnectorAttached;
-
-    private TextFieldValidationSupport validationSupport;
+    private TextFieldI18n i18n;
 
     private boolean manualValidationEnabled = false;
 
+    private String customErrorMessage;
+    private String constraintErrorMessage;
```

Comment:
What do you think about extracting this somehow? There could be a class that:
- Manages custom + constraint error message states
- Updates the effective error message property on the component when either is set

It wouldn't do a whole lot, but it looks like you have to replicate the logic quite a few times for different components. There's also `checkValidity` which is always the same though that one needs to access more stuff from the component. Anyway, maybe something to think about, we can also do this after updating all components and then see if there is something that works for all of them.

Reply:
Are you suggesting extracting this only for text fields or for all field components?

Reply:
I guess we could have something like `ErrorMessageController`:

```java
class ErrorMessageController {
  private Component component;

  public ErrorMessageController(HasValidation component) {
    this.component = component;
  }

  public void setCustomErrorMessage() {
    ...
    this.updateErrorMessageProperty();    
  }

  public void setConstraintErrorMessage() {
    ...
    this.updateErrorMessageProperty();
  }

  public void updateErrorMessageProperty() {
    this.component.getElement().setProperty("errorMessage");
  }
}
```

Reply:
> There's also checkValidity which is always the same though that one needs to access more stuff from the component. 

I discussed this with @yuriy-fix some time ago, and we concluded that it would currently complicate things especially since the whole i18n needs to be passed to that helper class or the class would need to provide an API to abstract itself from i18n. We agreed to revisit this in the future when we get to implementing error message providers.

Reply:
Yep, something like what you suggested with the controller. It seems like that would be applicable to all components, but not sure. But since there are other potential opportunities for code reuse, let's maybe finish changing the components first, and then see what we can extract as common logic?

Reply:
Agreed 👍